### PR TITLE
[**Part-1**] `neo-module/master`

### LINKED ISSUE #3230: Migrate Neo-Modules back to Neo core

### PR DESCRIPTION
**Pull request by cschuchardt88**
# Change Log
- Moved `neo-modules/src/` to directory `src`
- Moved `neo-modules/tests/` to directory `tests`
- Configured `.github` workflows for _tests_.
- Run `dotnet format`
- Added `neo-modules` projects to `neo.sln` file.

Closes #3230

image

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?
image

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

@shargon @Jim8y @vncoelho @superboyiii

### CODE REVIEW

**Comment by shargon** [comment]
what is part 2?

File: .editorconfig
Comment (shargon) [inline .editorconfig:30]: Why this change?

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:173]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:160]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:153]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:147]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:136]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:132]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:126]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:123]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:43]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:37]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:32]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:27]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:20]: Format change not related to the PR

```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:167]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:174]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:181]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:188]:
```suggestion
```

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:42]: `if: matrix.os != 'ubuntu-latest'` it's also required for `dotnet test` we need to do that always

File: .editorconfig
Comment (vncoelho) [reply .editorconfig:30]: @cschuchardt88,
 maybe it is better to no change anything now because my idea was to get the files and verify its integrity with a hash function in order to ensure no changes were done.

Comment (vncoelho) [reply .editorconfig:30]: @shargon, what is the best way to verify any change in any file?

**Review by vncoelho — COMMENTED** [review]
I see that you moved modules inside the root folder.
Why not src?

**Comment by cschuchardt88** [comment]
> I see that you moved modules inside the root folder. Why not src?

Because I figured it would make more sense to keep them separate.
1. Its unrelated to core development. _(their plugins)_
2. Easier to find and maintain.

> what is part 2?

Any of the commits above you don't want, would be part `2` or unrelated un-needed changes/fixes. Part `1` is just to make a working solution.

File: .github/workflows/main.yml
Comment (cschuchardt88) [reply .github/workflows/main.yml:42]: I removed the `Storage` tests from `windows` and `macos` this can be `permanent` or `temporarily`, up to you. Problem is that we would have to add `libleveldb` binary to this repo.

File: .editorconfig
Comment (cschuchardt88) [reply .editorconfig:30]: > Why this change?

Its a sample change, these files shouldn't have throughs at the end. But it can be removed.

> @shargon, what is the best way to verify any change in any file?

Tests? use a compare file tool?

**Comment by vncoelho** [comment]
@cschuchardt88,
I think it should be in the SRC in this first moment, like the others.

Furthermore, avoid any other change in files as we had discussed previously.
Thus, use any change in editorconfig in an additional PR.

**Comment by cschuchardt88** [comment]
At least with `plugins` name we limit yourself to only allowing `plugins` in this folder. But that can be done In the `sln` file to show. It can be big directory with `+13` items now being added. Becoming to hard to manage with unrelated items.

**Comment by cschuchardt88** [comment]
@vncoelho please look at the changes and let me know.

File: .github/workflows/main.yml
Comment (cschuchardt88) [reply .github/workflows/main.yml:42]: @shargon I think what I'm going to do is do what i was doing in #3018 with `leveldb` building to fix it for `windows` and `macos` along with `arm64`.

**Review by Jim8y — COMMENTED** [review]
You did not change the source file at all right? I need to compare the source files.

**Review by vncoelho — COMMENTED** [review]
@cschuchardt88,

it looks like everything is inside /src now.
I meant a folder inside /src...aheuaheuaea
I think you can create a folder named `Plugins` that will contain all these "modules/plugins".

There are too much projects now inside /src and this will not be good.

File: neo.sln
Comment (vncoelho) [inline neo.sln:43]: plugins?

**Comment by cschuchardt88** [comment]
> You did not change the source file at all right? I need to compare the source files.

I think one of them. `dotnet format` remove `this.` from a few files.

File: neo.sln
Comment (vncoelho) [reply neo.sln:43]: What is this @cschuchardt88 ?

**Review by vncoelho — COMMENTED** [review]
`[RpcClient](https://github.com/cschuchardt88/neo/tree/neo-module/master/src/RpcClient)` will not be moved?

**Comment by vncoelho** [comment]
I am not a user of RpcClient, so, maybe someone that uses it should recommend if it is a Plugin or another solution.

**Comment by vncoelho** [comment]
```cs
< 
<   <PropertyGroup>
<     <TargetFrameworks>net8.0</TargetFrameworks>
<     <PackageId>Neo.Plugins.OracleService</PackageId>
<   </PropertyGroup>
< 
<   <ItemGroup>
<     <PackageReference Include="NeoFS.API" Version="3.5.0" />
<   </ItemGroup>
< 
<   <ItemGroup>
<     <ProjectReference Include="..\..\Neo.ConsoleService\Neo.ConsoleService.csproj" />
<     <ProjectReference Include="..\RpcServer\RpcServer.csproj">
<       <Private>false</Private>
<       <ExcludeAssets>runtime</ExcludeAssets>
<     </ProjectReference>
<   </ItemGroup>
< 
---
> 	<PropertyGroup>
> 		<PackageId>Neo.Plugins.OracleService</PackageId>
> 	</PropertyGroup>
> 	<ItemGroup>
> 		<PackageReference Include="Neo.ConsoleService" Version="1.2.0" ExcludeAssets="runtime" />
> 		<PackageReference Include="NeoFS.API" Version="3.5.0" />
> 	</ItemGroup>
> 	<ItemGroup>
> 		<ProjectReference Include="..\RpcServer\RpcServer.csproj">
> 			<Private>false</Private>
> 			<ExcludeAssets>runtime</ExcludeAssets>
> 		</ProjectReference>
> 	</ItemGroup>
```

@cschuchardt88, let's not make any unnecessary changes, as we already discussed before and here.

For example, in this case you modified the reference of `Neo.ConsoleService` from local to a nuget.
This was not agreed and this is not the purpose of this PR.

I am verifying the changes of each file now in order to ensure that there are no changes.
This PR is just necessary changes for the plugins to be moved.

**Comment by vncoelho** [comment]
```cs
/neo-modules/src//ApplicationLogs/ApplicationLogs.csproj
3d2
<     <TargetFrameworks>net8.0</TargetFrameworks>
10c9
<     <ProjectReference Include="..\..\Neo.ConsoleService\Neo.ConsoleService.csproj" />
---
>     <ProjectReference Include="..\..\neo\src\Neo.ConsoleService\Neo.ConsoleService.csproj" />
diff /home/vitor/Github/neo-tests/docker-build-neo-cli/localfiles/neo-cris/src/Plugins//ApplicationLogs/Settings.cs /home/vitor/Github/neo-tests/docker-build-neo-cli/localfiles/neo-modules/src//ApplicationLogs/Settings.cs
28,31c28,31
<             Path = section.GetValue("Path", "ApplicationLogs_{0}");
<             Network = section.GetValue("Network", 5195086u);
<             MaxStackSize = section.GetValue("MaxStackSize", (int)ushort.MaxValue);
<             Debug = section.GetValue("Debug", false);
---
>             this.Path = section.GetValue("Path", "ApplicationLogs_{0}");
>             this.Network = section.GetValue("Network", 5195086u);
>             this.MaxStackSize = section.GetValue("MaxStackSize", (int)ushort.MaxValue);
>             this.Debug = section.GetValue("Debug", false);

```

For the application log you kept the local reference.

**Comment by vncoelho** [comment]
```cs
< 
<   <PropertyGroup>
<     <TargetFrameworks>net8.0</TargetFrameworks>
<     <PackageId>Neo.Plugins.TokensTracker</PackageId>
<   </PropertyGroup>
< 
<   <ItemGroup>
<     <ProjectReference Include="..\RpcServer\RpcServer.csproj" />
<   </ItemGroup>
< 
---
> 	<PropertyGroup>
> 		<PackageId>Neo.Plugins.TokensTracker</PackageId>
> 	</PropertyGroup>
> 	<ItemGroup>
> 		<ProjectReference Include="..\RpcServer\RpcServer.csproj">
> 			<Private>false</Private>
> 			<ExcludeAssets>runtime</ExcludeAssets>
> 		</ProjectReference>
> 	</ItemGroup>
```

If possible remove `<ExcludeAssets>runtime</ExcludeAssets>` and `<Private>false</Private>`.
Add this in another PR.

**Review by vncoelho — COMMENTED** [review]
@cschuchardt88,

Please fix my last comments.
* Fix Console.Service reference to local where you changed them
* Remove the additional parameter on TokensTracker plugin .csproj
* Move RpcClient to Plugins as well (after the merge you create another one to move to src). I think it will be better

**Comment by vncoelho** [comment]
@cschuchardt88,
after this last changes I mentioned are attended, what do you need from my side?
Is everything already compiling and running? Where you able to do some tests?
How about coverage?

File: .editorconfig
Comment (vncoelho) [inline .editorconfig:308]: Revert the changes here.

File: .gitattributes
Comment (vncoelho) [inline .gitattributes:7]: Is there any .props file in the plugins being merged?

File: neo.sln
Comment (cschuchardt88) [reply neo.sln:43]: The folders inside the solution. Check _screenshot_ to see.

**Comment by cschuchardt88** [comment]
> I am not a user of RpcClient, so, maybe someone that uses it should recommend if it is a Plugin or another solution.

This is not a plugin. It's a `dll` for `3rd` party applications to use.

> @cschuchardt88, let's not make any unnecessary changes, as we already discussed before and here.
> 
> For example, in this case you modified the reference of `Neo.ConsoleService` from local to a nuget. This was not agreed and this is not the purpose of this PR.
> 
> I am verifying the changes of each file now in order to ensure that there are no changes. This PR is just necessary changes for the plugins to be moved.

These are all the required projects to make the application build. `dotnet` adds these when we include a project from the solution. I must of missed that one from `nuget`.

> If possible remove `<ExcludeAssets>runtime</ExcludeAssets>` and `<Private>false</Private>`. Add this in another PR.

Again `dotnet` adds this, but there is different choices you can choose from. Another PR would be best.

> @cschuchardt88, after this last changes I mentioned are attended, what do you need from my side? Is everything already compiling and running? Where you able to do some tests? How about coverage?

It's all working well. Best test is `dotnet test` or `github` workflow.

File: .editorconfig
Comment (cschuchardt88) [reply .editorconfig:308]: what was here?

File: .gitattributes
Comment (cschuchardt88) [reply .gitattributes:7]: yes have 3 different ones

**Comment by cschuchardt88** [comment]
@shargon some input would be nice on the purpose and integrity this PR and repo move.

**Comment by vncoelho** [comment]
@cschuchardt88 ,please attend the previous comment for us to move forward.

**Comment by vncoelho** [comment]
Console service to local please.

**Comment by cschuchardt88** [comment]
> Console service to local not nuget, later you create a PR on that if that is the case.

I got confused, thought you wanted me to change that to `nuget`. Should be fixed now.

File: .github/workflows/main.yml
Comment (shargon) [reply .github/workflows/main.yml:42]: We need to test the storage

Comment (cschuchardt88) [reply .github/workflows/main.yml:42]: @shargon what is a solution for `libleveldb` binaries in repo?

Comment (shargon) [reply .github/workflows/main.yml:42]: This can be discussed in a different pr, this should be only a move

Comment (cschuchardt88) [reply .github/workflows/main.yml:42]: Agree and I'll issue to for a note to fix it.

File: .github/workflows/main.yml
Comment (shargon) [inline .github/workflows/main.yml:37]:
```suggestion
        sudo apt-get --assume-yes install libleveldb-dev librocksdb-dev
```

File: .github/workflows/main.yml
Comment (cschuchardt88) [inline .github/workflows/main.yml:37]:
```suggestion
        dotnet sln neo.sln remove ./tests/Neo.Plugins.Storage.Tests/Neo.Plugins.Storage.Tests.csproj
        sudo apt-get --assume-yes install libleveldb-dev librocksdb-dev
```

File: .github/workflows/main.yml
Comment (vncoelho) [inline .github/workflows/main.yml:35]: On the neo-modules repo, tests were running just on " ubuntu-latest".
Since coverage is just on ubuntu-latests as well, I think that if there is any trouble we can push the fix in another PR.

**Comment by vncoelho** [comment]
In principle, checking the logs, it looks like everything was fine.
However, the coverage summary is not directly appearing on the PR.

image

**Comment by cschuchardt88** [comment]
> In principle, checking the logs, it looks like everything was fine. However, the coverage summary is not directly appearing on the PR.
> 
> image

I believe if it doesn't than coverage didn't change.

**Comment by vncoelho** [comment]
For Plugins the coverage is correct.
"coverage/coveralls (push) - Coverage remained the same at 48.583%" -> This is from the last commit on neo-modules

**Comment by vncoelho** [comment]
The other also increased a little. I think it is all correct.
We just have some display problems here in the PR. We can fix later if they persist.

**Review by vncoelho — COMMENTED** [review]
I am adjusting my tooling and will test in some minutes.

**Review by vncoelho — COMMENTED** [review]
```cs
#36 71.70 /usr/share/dotnet/sdk/8.0.203/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(112,5): error NETSDK1152: Found multiple publish output files with the same relative path: /opt/neoLib/src/Plugins/RpcServer/config.json, /opt/neoLib/src/Plugins/StateService/config.json. [/opt/neoLib/src/Plugins/StateService/StateService.csproj::TargetFramework=net8.0]
```

**Comment by vncoelho** [comment]
For me it is ready to merge.

Code is the same as neo-modules apart from necessary changes.
Compilation, tests and publish are all good in my local environment.

However, code is not running due to recent PRs that introduces some problems on wallet.
I think we can merge and then fix in order to move forward.

**Comment by Jim8y** [comment]
If code is good, we can go with this pr already. Not necessarily to complete everything in one pr. But maybe we need a development branch? @shargon

**Comment by cschuchardt88** [comment]
I have #3244 for changes

**Review by Jim8y — APPROVED** [review]

**Comment by vncoelho** [comment]
> If code is good, we can go with this pr already. Not necessarily to complete everything in one pr. But maybe we need a development branch? @shargon

If this please goes to a development branch please revert the last 3 commits as well and migrate then to such branch.

**Comment by shargon** [comment]
Give me a couple of hours for the comparison

**Comment by vncoelho** [comment]
@Jim8y , I think we can try with the master branch a little bit more now that we have "all necessary components" here.
I think that by pushing a UT that tests node's ability to start a consensus and do basic operations will be very good. Maybe we can keep working on main branch in such case.

But I am not against using a dev branch.

**Review by roman-khimov — COMMENTED** [review]
Code-wise (at 68e4146f6d5633e670de009125167ef672e439e8) I've noticed some `s/this//` changes and two lines moved in `src/Plugins/RpcClient/Models/RpcInvokeResult.cs` compared to `master` of `neo-modules`. Minor, but can be fixed to have cleaner initial state here and stylistic changes in separate PRs.

Can't comment on csproj/workflows changes, likely OK.

**Review by vncoelho — CHANGES_REQUESTED** [review]
Perhaps the aforementioned problem I mentioned on discord was not introduced on [08f2dfc56762bb43be33e873bc3659bad368d4d6](https://github.com/neo-project/neo/pull/3204), but here.

In this case, please do not merge until I finish my tests because the error here may take time to solve otherwise.

**Review by vncoelho — APPROVED** [review]
The error was, in fact, a typo on the new configuration file name on Part-2 PR.
Thus, it is approved now with tests running as expected.

**Pull request merged**
